README.md
# SUPER NOVA

Super Nova is a hand eye coordination focused arcade game.

  
  

# GAMEPLAY

The **objective** of the game is to hit the most amount of falling **Disks** as accurately as possible with either the '***A***', '***S***' or '***D***' key. Those keys are your **Detonators**.

  

## Timer

There will be a time limit of ***100*** seconds.

  

> So make every second count!

  

## Score

You will be rewarded various **Score Points** for each hit based off of timing and accuracy.

  

## Gravity Level

Your **Gravity Level** determines how quickly the disks spawn and fall. As you hit disks, you will be rewarded **Skill Rating Points** behind the scenes which contribute towards promoting your gravity level. If you miss **Disks** or allow them to pass by then you will lose skill rating and possibly demote your gravity level. The highest gravity level achievable is ***10***.

  

## Streak Count

You can grow a **Streak Count** for hitting multiple **Disks** in a row without missing.

  

## Score Multiplier

You have a **Score Multiplier** which can double, triple or even quadruple the amount of **Score Points** you earn for each hit. The higher your **Streak Count**, the more your score multiplier grows with ***x4*** being the max.

  

## Missing A Disk

If you happen to miss a **Disk** or trigger an empty **Detonator** then your **Streak Count** and **Score Multiplier** bonus will both reset. Build it back up to earn as many **Score Points** as possible.

  

## Power Bar Bonus Mode

Hitting an **Accuracy Rating** of '***Perfect***' or '***Super Perfect***' grants you **Power Bar** points for it's respective column up to 100 points. Once you hit the cap your power bar will start to shrink and will spawn special 'Bonus Disks'. Hitting a Bonus Disk grants you 300 points regardless of accuracy (That's triple the max of a regular disk). They do not grant power bar points.

# TECHNOLOGIES

 **HTML**
 **CSS**
 **JavaScript**

# ISSUES

- **Settings** panel has not been created yet.
- **Timer** does not continue when game is tabbed out.

# EASTER EGG

  

> Press the '***E***' key anytime to make a wish ;)

  

##

  
  

### *Best of luck,*

#### *Novasharx*

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2023-03-15

### Added

- `CHANGELOG.md` file.

### Fixed
- Updated `README.md`file to be formatted properly.

index.js
const canvas = document.querySelector('canvas')
const ctx = canvas.getContext('2d')

canvas.width = 900
canvas.height = 900

const refreshRate = 60 // fps

const gameManager = new GameManager()
const gameTimer = new Timer()

var diskAssembly = {
    red: [],
    green: [],
    blue: [],
    hitDisks: [],
    missedDisks: []
}

//red, green, blue
const triColors = ['#DB3324', '#24DB33', '#3324DB']

const gameplayDetails = {
    position: {x: null, y: 150},
    runwayWidth: 150,
    runwayHeight: 500,
    runwayOffset: 10
}

const runwayRed = new Runway({
    position: {
        x: (canvas.width/2) - (gameplayDetails.runwayWidth * 1.5) - gameplayDetails.runwayOffset,
        y: gameplayDetails.position.y
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayHeight,
    color: triColors[0]
})
const runwayGreen = new Runway({
    position: {
        x: runwayRed.position.x + gameplayDetails.runwayWidth + gameplayDetails.runwayOffset,
        y: gameplayDetails.position.y
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayHeight,
    color: triColors[1]
})
const runwayBlue = new Runway({
    position: {
        x: runwayGreen.position.x + gameplayDetails.runwayWidth + gameplayDetails.runwayOffset,
        y: gameplayDetails.position.y
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayHeight,
    color: triColors[2]
})

const detonatorRed = new Detonator({
    position: {
        x: runwayRed.position.x,
        y: runwayRed.position.y + gameplayDetails.runwayHeight + gameplayDetails.runwayOffset
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayWidth,
    color: triColors[0],
    imageSrc: './Assets/Images/Detonator_red.png'
})
const detonatorGreen = new Detonator({
    position: {
        x: runwayGreen.position.x,
        y: runwayGreen.position.y + gameplayDetails.runwayHeight + gameplayDetails.runwayOffset
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayWidth,
    color: triColors[1],
    imageSrc: './Assets/Images/Detonator_green.png'
})
const detonatorBlue = new Detonator({
    position: {
        x: runwayBlue.position.x,
        y: runwayBlue.position.y + gameplayDetails.runwayHeight + gameplayDetails.runwayOffset
    },
    width: gameplayDetails.runwayWidth,
    height: gameplayDetails.runwayWidth,
    color: triColors[2],
    imageSrc: './Assets/Images/Detonator_blue.png'
})

function animate() {
    setTimeout(function () { window.requestAnimationFrame(animate) }, 1000 / refreshRate)

    ctx.fillStyle = 'black'
    ctx.fillRect(0, 0, canvas.width, canvas.height)

    runwayRed.update()
    runwayGreen.update()
    runwayBlue.update()

    detonatorRed.update()
    detonatorGreen.update()
    detonatorBlue.update()

    diskAssembly.red.forEach((disk) => {
        disk.update()
    })
    diskAssembly.green.forEach((disk) => {
        disk.update()
    })
    diskAssembly.blue.forEach((disk) => {
        disk.update()
    })
    diskAssembly.hitDisks.forEach((disk) => {
        disk.update()
    })
    diskAssembly.missedDisks.forEach((disk) => {
        disk.update()
    })

    renderSpawningMask()

    gameTimer.render()
}

animate()

utility.js
const shootingStarDisplay = document.getElementById('shooting-star')

const devModeButton = document.getElementById('dev-mode')

const mainMenuDisplay = document.getElementById('main-menu-container')
const tutorialButton = document.getElementById('tutorial')
const tutorialDisplay = document.getElementById('tutorial-display')
const tutorialImage = document.getElementById('tutorial-image')
const tutorialDescription = document.getElementById('tutorial-description')
const tutorialNextButton = document.getElementById('next-tutorial')
const tutorialPrevButton = document.getElementById('prev-tutorial')
const tutorialCloseButton = document.getElementById('close-tutorial')
const playButton = document.getElementById('play-button')
const settingsButton = document.getElementById('settings')

const inGameDisplay = document.getElementById('in-game-container')
const scoreDisplay = document.getElementById('score')
const streakDisplay = document.getElementById('streak')
const scoreMultiplierDisplay = document.getElementById('scoremultiplier')
const gravityLevelDisplay = document.getElementById('gravity')

const postGameDisplay = document.getElementById('post-game-container')
const finalScoreDisplay = document.getElementById('stat-final-score')
const highestGravityDisplay = document.getElementById('stat-highest-gravity')
const longestStreakDisplay = document.getElementById('stat-longest-streak')
const disksHitDisplay = document.getElementById('stat-disks-hit')
const averageAccuracyDisplay = document.getElementById('stat-average-accuracy')
const playAgainButton = document.getElementById('play-again-button')
const backToMenuButton = document.getElementById('back-to-menu')

const hitAudio = new Audio('./Assets/Audio/HitShort.wav')
const missAudio = new Audio('./Assets/Audio/Miss.wav')
const logoSwirlAudio = new Audio ('./Assets/Audio/Logo_SwirlShort.wav')

function renderSpawningMask() {
    ctx.fillStyle = 'black'
    ctx.fillRect(runwayRed.position.x - 3, 0, gameplayDetails.runwayWidth * 3 + gameplayDetails.runwayOffset * 3, gameplayDetails.position.y - 1)
}

function spawnDisk(id) {
    let columnPosX
    let runwayBonusState = false

    switch (id) {
        case 'red':
            columnPosX = runwayRed.position.x
            if (runwayRed.bonusState) {
                runwayBonusState = true
            }
            break;
        case 'green':
            columnPosX = runwayGreen.position.x
            if (runwayGreen.bonusState) {
                runwayBonusState = true
            }
            break;
        case 'blue':
            columnPosX = runwayBlue.position.x
            if (runwayBlue.bonusState) {
                runwayBonusState = true
            }
            break;
    }

    let disk = new Disk({
        position: {
            x: columnPosX,
            y: gameplayDetails.position.y - 150
        },
        imageSrc: `./Assets/Images/Disk_${id}.png`,
        bonusState: runwayBonusState,
        id: id
    })

    if (disk.bonusState) {
        disk.image.src = `./Assets/Images/Disk_Bonus_${id}.png`
    }

    switch (id) {
        case 'red':
            diskAssembly.red.push(disk)
            break;
        case 'green':
            diskAssembly.green.push(disk)
            break;
        case 'blue':
            diskAssembly.blue.push(disk)
            break;
    }
    gameManager.disksSpawned++
}

function runwayDiskSpawner(id) {
    setTimeout(() => {
        spawnDisk(id)
        runwayDiskSpawner(id)
    }, 4000 / gameManager.gravity);
}

function randomDiskSpawner() {
    gameManager.diskSpawnerId = setTimeout(() => {
        let random = Math.floor(Math.random() * 3)
        let id
        switch (random) {
            case 0:
                id = 'red'
                break;
            case 1:
                id = 'green'
                break;
            case 2:
                id = 'blue'
                break;
        }
        spawnDisk(id)
        randomDiskSpawner()
    }, 4000 / gameManager.gravity)
}

function diskMissed(id) {
    switch (id) {
        case 'red':
            detonatorRed.image.src = './Assets/Images/Detonator_Miss_red.png'
            break;
        case 'green':
            detonatorGreen.image.src = './Assets/Images/Detonator_Miss_green.png'
            break;
        case 'blue':
            detonatorBlue.image.src = './Assets/Images/Detonator_Miss_blue.png'
            break;
    }

    gameManager.updateSkillRating('lose')
    gameManager.updateStreak('miss')

    missAudio.load()
    missAudio.play()
}

function spawnShootingStar() {
    shootingStarDisplay.src = './Assets/Images/Shooting_Star_07.png'
    let position = {
        x: Math.floor(Math.random() * (window.innerWidth - 150) + 1),
        y: Math.floor(Math.random() * (window.innerHeight - 150) + 1)
    }
    setTimeout(() => {
        shootingStarDisplay.style.top = `${position.y}px`
        shootingStarDisplay.style.left = `${position.x}px`
        shootingStarDisplay.src = './Assets/Images/Shooting_Star.gif'
    }, 0)
}

// Generate a shooting star between minTIme and maxTime perpetually
function randomShootingStars() {
    let minTime = 30000
    let maxTime = 600000
    let randomTime = Math.floor((Math.random() * (maxTime - minTime + 1)) + minTime)
    setTimeout(() => {
        spawnShootingStar()
        randomShootingStars()
    }, randomTime)
}

function promptDevMode() {
    response = window.prompt('Please ENTER the Password')
    if (response === 'nova') {
        gameManager.devMode = true
        devModeButton.style.color = 'white'
        window.alert('Success!')
    } else {
        window.alert('Denied!')
    }
}

window.addEventListener('keypress', (event) => {
    if (gameManager.gameState === 'Active') {
        switch (event.key) {

            // Regular Game Buttons
            case 'a':
                if (diskAssembly.red.length === 0) {
                    diskMissed('red')
                } else {
                    diskAssembly.red[0].detonate()
                }
                break;
            case 's':
                if (diskAssembly.green.length === 0) {
                    diskMissed('green')
                } else {
                    diskAssembly.green[0].detonate()
                }
                break;
            case 'd':
                if (diskAssembly.blue.length === 0) {
                    diskMissed('blue')
                } else {
                    diskAssembly.blue[0].detonate()
                }
                break;
        }
    }

    if (gameManager.gameState === 'Active' && gameManager.devMode) {
        switch (event.key) {

            // Game Dev buttons
        case '1':
            spawnDisk('red')
            break;
        case '2':
            spawnDisk('green')
            break;
        case '3':
            spawnDisk('blue')
            break;
        case '4':
            runwayDiskSpawner('red')
            break;
        case '5':
            runwayDiskSpawner('green')
            break;
        case '6':
            runwayDiskSpawner('blue')
            break;
        case '0':
            randomDiskSpawner()
            break;
        case ' ':
            gameManager.startGame()
            break;

        }
    }

    switch (event.key) {
        case 'e':
            spawnShootingStar()
            break;
    }
})

window.addEventListener('keyup', (event) => {
    switch (event.key) {
        case 'a':
            detonatorRed.image.src = './Assets/Images/Detonator_red.png'
            break;
        case 's':
            detonatorGreen.image.src = './Assets/Images/Detonator_green.png'
            break;
        case 'd':
            detonatorBlue.image.src = './Assets/Images/Detonator_blue.png'
            break;
    }
})

devModeButton.addEventListener('click', () => {
    if (!gameManager.devMode) {
        promptDevMode()
    } else {
        gameManager.devMode = false
        devModeButton.style.color = 'rgb(129, 129, 129)'
    }
})

tutorialButton.addEventListener('click', () => {
    tutorialDisplay.style.display = 'flex'
    gameManager.gameState = 'Menu-Tutorial'
    gameManager.tutorialSlide = 0
    tutorialImage.src = `./Assets/Images/tutorial_slide_${gameManager.tutorialSlide}.jpg`
    tutorialDescription.innerHTML = gameManager.tutorialDescriptions[gameManager.tutorialSlide]
})

tutorialNextButton.addEventListener('click', () => {
    gameManager.tutorialSlide++
    tutorialImage.src = `./Assets/Images/tutorial_slide_${gameManager.tutorialSlide}.jpg`
    tutorialDescription.innerHTML = gameManager.tutorialDescriptions[gameManager.tutorialSlide]
})

tutorialPrevButton.addEventListener('click', () => {
    gameManager.tutorialSlide--
    tutorialImage.src = `./Assets/Images/tutorial_slide_${gameManager.tutorialSlide}.jpg`
    tutorialDescription.innerHTML = gameManager.tutorialDescriptions[gameManager.tutorialSlide]
})

tutorialCloseButton.addEventListener('click', () => {
    tutorialDisplay.style.display = 'none'
    gameManager.gameState = 'Main-Menu'
})

playButton.addEventListener('click', () => {
    gameManager.startGame()
})

playAgainButton.addEventListener('click', () => {
    gameManager.startGame()
})

backToMenuButton.addEventListener('click', () => {
    gameManager.loadMainMenu()
})

src/classes.ts
interface position {
    x: number,
    y: number
}

interface SpriteProperties {
    position: position,
    width: number,
    height: number,
    imageSrc: string,
}

interface ColoredSpriteProperties extends SpriteProperties {
    color: string
}

interface DiskSpriteProperties extends SpriteProperties {
    id: string,
    bonusState: boolean,
}

class Sprite {

    position: position
    width: number
    height: number
    image: HTMLImageElement

    constructor({
        position,
        width,
        height,
        imageSrc,
    }: SpriteProperties) {
        this.position = position
        this.width = width
        this.height = height
        this.image = new Image()
        // this.image.src = imageSrc
    }

    draw() {
        //image to be drawn
        ctx.drawImage(
            this.image,
            //coordinates to drawn
            this.position.x, this.position.y,
            //width and height
            this.width, this.height
        )
    }
}

class Runway extends Sprite {

    color: string | undefined
    strokeStrength: number
    powerBar: number
    bonusState: boolean
    bonusDurationID: number | undefined

    constructor({
        position,
        width,
        height,
        color,
        imageSrc
    }: ColoredSpriteProperties) {
        super({
            position,
            width,
            height,
            imageSrc
        })

        this.image = new Image()
        // this.image.src = imageSrc
        this.color = color
        this.strokeStrength = 2
        this.powerBar = 0
        this.bonusState = false
        this.bonusDurationID
    }

    render() {
        ctx.lineWidth = this.strokeStrength
        ctx.fillStyle = this.color + '50'
        ctx.fillRect(this.position.x, this.position.y, this.width, this.height * (this.powerBar / 100))
        ctx.strokeStyle = this.color
        ctx.strokeRect(this.position.x, this.position.y, this.width, this.height)
    }

    update() {
        if (this.powerBar >= 100 && !this.bonusState) {
            this.activatBonusState()
        }
        this.render()
    }

    activatBonusState() {
        this.bonusState = true
        this.strokeStrength = 5
        this.bonusDurationID = setInterval(() => {
            this.powerBar -= 0.1
            if (this.powerBar <= 0) {
                this.powerBar = 0
                this.deactivateBonusState()
            }
        }, 10)
    }

    deactivateBonusState() {
        clearInterval(this.bonusDurationID)
        this.bonusState = false
        this.strokeStrength = 2
    }
}

class Detonator extends Sprite {

    color: string | undefined

    constructor({
        position,
        width,
        height,
        color,
        imageSrc
    }: ColoredSpriteProperties) {
        super({
            position,
            width,
            height,
            imageSrc
        })

        this.image = new Image()
        this.image.src = imageSrc
        this.color = color
    }

    update() {
        this.draw()
    }

}

class Disk extends Sprite {
    hasPassed: boolean;
    gotErased: boolean;
    id: string;
    gotHit: boolean;
    bonusState: boolean;
    constructor({
        position,
        width,
        height,
        imageSrc,
        bonusState,
        id
    }: DiskSpriteProperties) {
        super({
            position,
            width,
            height,
            imageSrc,
        })

        this.image = new Image()
        this.image.src = imageSrc
        this.width = 150
        this.height = 150
        this.hasPassed = false
        this.gotErased = false
        this.gotHit = false
        this.bonusState = bonusState
        this.id = id
    }

    update() {
        this.draw()
        this.position.y += gameManager.gravity

        if (!this.hasPassed &&
            !this.gotHit &&
            this.position.y + 25 > detonatorRed.position.y + detonatorRed.height - 25) {
            this.diskHasPassed()
        }

        if (!this.gotErased && this.position.y > canvas.height) {
            this.eraseDisk()
        }
    }

    detonate() {
        if (!this.gotHit &&
            this.position.y - 50 + this.height > detonatorRed.position.y &&
            this.position.y + 50 < detonatorRed.position.y + detonatorRed.height) {
            this.diskGotHit()
        } else {
            diskMissed(this.id)
        }
    }

    diskGotHit() {
        gameManager.diskHitCount++
        hitAudio.load()
        hitAudio.play()
        gameManager.updateSkillRating('gain')
        gameManager.updateStreak('hit')
        this.processAccuracy()
        this.gotHit = true
        let target: Disk | undefined
        setTimeout(() => {
            switch (this.id) {
                case 'red':
                    target = diskAssembly.red.shift()
                    diskAssembly.hitDisks.push(target)
                    detonatorRed.image.src = './Assets/Images/Detonator_Hit_red.png'
                    break;
                case 'green':
                    target = diskAssembly.green.shift()
                    diskAssembly.hitDisks.push(target)
                    detonatorGreen.image.src = './Assets/Images/Detonator_Hit_green.png'
                    break;
                case 'blue':
                    target = diskAssembly.blue.shift()
                    diskAssembly.hitDisks.push(target)
                    detonatorBlue.image.src = './Assets/Images/Detonator_Hit_blue.png'
                    break;
            }
        }, 0)
    }

    processAccuracy() {
        if (this.bonusState) {
            this.image.src = `./Assets/Images/Accuracy_BONUS_${this.id}.png`
            gameManager.updateScore(300)
        } else {

            let offset = Math.abs(this.position.y - detonatorRed.position.y)
            let percentage = Math.round(((95 - offset) / 95) * 100)

            gameManager.accuracySum += percentage
            gameManager.averageAccuracy = Math.round(gameManager.accuracySum / gameManager.diskHitCount)

            if (percentage >= 93) {
                this.image.src = `./Assets/Images/Accuracy_SUPERPERFECT_${this.id}.png`
                gameManager.increasePowerBar(this.id, 5)
                gameManager.updateScore(100)
            }
            else if (percentage >= 80) {
                this.image.src = `./Assets/Images/Accuracy_PERFECT_${this.id}.png`
                gameManager.increasePowerBar(this.id, 2)
                gameManager.updateScore(65)
            }
            else if (percentage >= 50) {
                this.image.src = `./Assets/Images/Accuracy_GOOD.png`
                gameManager.updateScore(25)
            }
            else {
                this.image.src = `./Assets/Images/Accuracy_MEH.png`
                gameManager.updateScore(10)
            }
        }
    }

    diskHasPassed() {
        gameManager.updateSkillRating('lose')
        gameManager.updateStreak('miss')
        this.hasPassed = true
        if (!this.gotHit) {
            this.image.src = `./Assets/Images/Missed_Disk_${this.id}.png`
        }
        let target: Disk
        setTimeout(() => {
            switch (this.id) {
                case 'red':
                    target = diskAssembly.red.shift()
                    diskAssembly.missedDisks.push(target)
                    break;
                case 'green':
                    target = diskAssembly.green.shift()
                    diskAssembly.missedDisks.push(target)
                    break;
                case 'blue':
                    target = diskAssembly.blue.shift()
                    diskAssembly.missedDisks.push(target)
                    break;
            }
        }, 0)
    }

    eraseDisk() {
        this.gotErased = true
        setTimeout(() => {
            if (this.gotHit) {
                diskAssembly.hitDisks.shift()
            } else {
                diskAssembly.missedDisks.shift()
            }
        }, 0);
    }
}

class Timer {

    currentTimerId: number | undefined
    currentTime: number
    maxTime: number
    timeRemainingFraction: number

    constructor() {
        this.currentTimerId
        this.currentTime = 0
        this.maxTime = 100
        this.timeRemainingFraction = this.currentTime / this.maxTime
    }

    render() {
        ctx.strokeStyle = 'rgba(255, 255, 255, 0.2)'
        ctx.lineWidth = 5
        ctx.beginPath()
        ctx.arc(canvas.width / 2, canvas.height / 2, canvas.width / 2 - 15, Math.PI * 0.5, Math.PI * ((2 * this.timeRemainingFraction) + 0.5))
        ctx.stroke()
    }

    beginTimer() {
        this.currentTime = this.maxTime
        this.timeRemainingFraction = this.currentTime / this.maxTime
        this.currentTimerId = setInterval(() => {
            this.currentTime -= 0.01
            if (this.currentTime <= 0) {
                this.endTimer()
                gameManager.endGame()
            } else {
                this.timeRemainingFraction = this.currentTime / this.maxTime
            }
        }, 10);
    }

    endTimer() {
        clearInterval(this.currentTimerId)
        this.currentTime = 0
        this.timeRemainingFraction = this.currentTime / this.maxTime
    }
}

class GameManager {
    constructor() {
        this.devMode = false
        this.gameState = 'Main-Menu'
        this.tutorialSlide = 1
        this.tutorialDescriptions = [
            `OBJECTIVE <br><br> The objective of the game is to hit the most amount of falling disks as accurately as possible with either the 'A', 'S' or 'D' key. Those keys are your detonators.`,
            `TIMER <br><br> There will be a time limit of 100 seconds. So make every second count!`,
            `SCORE <br><br> You will be rewarded various score points for each hit based off of timing and accuracy.`,
            `GRAVITY LEVEL <br><br> Your 'Gravity Level' determines how quickly the disks spawn and fall. As you hit disks, you will be rewarded skill rating points behind the scenes which contribute towards promoting your Gravity Level. If you miss disks or allow them to pass by then you will lose skill rating and possibly demote your Gravity Level. The highest gravity level is 10.`,
            `STREAK COUNT <br><br> You can grow a streak for hitting multiple disks in a row without missing.`,
            `SCORE MULTIPLIER <br><br> You have a score multiplier which can double, triple or even quadrople the amount of points you earn for each hit. The higher your streak count, more your multiplier grows with 'x4' being the max.`,
            `MISSING A DISK <br><br> If you happen to miss a disk or trigger an empty detonator then your streak count and score multiplier bonus will both reset. Build it back up to earn as many points as possible.`,
            `POWER BAR BONUS MODE <br><br> Hitting an accuracy rating of 'Perfect' or 'Super Perfect' grants you powerbar points for it's respective column up to 100 points. Once you hit the cap your power bar will start to shrink and will spawn special 'Bonus Disks'. Hitting a Bonus Disk grants you 300 points regardless of accuracy (That's triple the max of a regular disk). They do not grant power bar points.`,
            `EASTER EGG <br><br> Press the 'E' key to make a wish ; )`
        ]
        this.diskSpawnerId = null
        this.gravityLvl = 1
        this.skillRating = 0
        this.gravity = 4 + this.gravityLvl //gravity should be from 5 to 15 i.e lvl 1 to 10
        this.currentScore = 0
        this.streakCounter = 0
        this.scoreMultiplier = 1
        this.disksSpawned = 0
        this.diskHitCount = 0
        this.accuracySum = 0
        this.averageAccuracy = 0
        this.highestGravity = 1
        this.longestStreak = 0
    }

    resetGameValues() {
        this.diskSpawnerId = null
        this.gravityLvl = 1
        this.gravity = 4 + this.gravityLvl
        this.skillRating = 0
        this.currentScore = 0
        this.streakCounter = 0
        this.scoreMultiplier = 1
        this.disksSpawned = 0
        this.diskHitCount = 0
        this.accuracySum = 0
        this.averageAccuracy = 0
        this.highestGravity = 1
        this.longestStreak = 0
        diskAssembly = {
            red: [],
            green: [],
            blue: [],
            hitDisks: [],
            missedDisks: []
        }
        runwayRed.powerBar = 0
        runwayGreen.powerBar = 0
        runwayBlue.powerBar = 0
        runwayRed.deactivateBonusState()
        runwayGreen.deactivateBonusState()
        runwayBlue.deactivateBonusState()
        gravityLevelDisplay.innerHTML = this.gravityLvl
        scoreDisplay.innerHTML = this.currentScore
        streakDisplay.innerHTML = this.streakCounter
        scoreMultiplierDisplay.innerHTML = this.scoreMultiplier
    }

    startGame() {
        this.resetGameValues()
        this.gameState = 'Active'
        mainMenuDisplay.style.display = 'none'
        postGameDisplay.style.display = 'none'
        inGameDisplay.style.display = 'block'
        canvas.style.transition = '1s'
        canvas.style.boxShadow = `0px 0px 100px rgba(255, 255, 255, 0.1)`
        gameTimer.beginTimer()
        randomDiskSpawner()
    }

    updateScore(score) {
        this.currentScore += (score * this.scoreMultiplier)
        scoreDisplay.innerHTML = this.currentScore
    }

    updateStreak(value) {
        switch (value) {
            case 'miss':
                this.streakCounter = 0
                streakDisplay.style.animation = 'none'
                setTimeout(() => {
                    streakDisplay.style.animation = '0.2s streakReset'
                }, 0)
                break;
            case 'hit':
                this.streakCounter++
                streakDisplay.style.animation = 'none'
                setTimeout(() => {
                    streakDisplay.style.animation = '0.2s pulse'
                }, 0)
                break;
        }
        this.updateMultiplier()

        if (this.longestStreak < this.streakCounter) {
            this.longestStreak = this.streakCounter
        }

        streakDisplay.innerHTML = this.streakCounter
    }

    updateMultiplier() {
        if (this.streakCounter >= 25) {
            if (this.scoreMultiplier != 4) {
                scoreMultiplierDisplay.style.animation = 'none'
                setTimeout(() => {
                    scoreMultiplierDisplay.style.animation = '0.2s pulse'
                })
            }

            scoreMultiplierDisplay.style.fontSize = '2.5em'
            this.scoreMultiplier = 4
        }
        else if (this.streakCounter >= 15) {
            if (this.scoreMultiplier != 3) {
                scoreMultiplierDisplay.style.animation = 'none'
                setTimeout(() => {
                    scoreMultiplierDisplay.style.animation = '0.2s pulse'
                })
            }

            scoreMultiplierDisplay.style.fontSize = '2.2em'
            this.scoreMultiplier = 3
        }
        else if (this.streakCounter >= 5) {
            if (this.scoreMultiplier != 2) {
                scoreMultiplierDisplay.style.animation = 'none'
                setTimeout(() => {
                    scoreMultiplierDisplay.style.animation = '0.2s pulse'
                })
            }

            scoreMultiplierDisplay.style.fontSize = '1.8em'
            this.scoreMultiplier = 2
        } else {
            this.scoreMultiplier = 1
            scoreMultiplierDisplay.style.fontSize = '1.5em'
        }

        scoreMultiplierDisplay.innerHTML = `x${this.scoreMultiplier}`
    }

    increasePowerBar(id, power) {
        switch (id) {
            case 'red':
                if (!runwayRed.bonusState) {
                    runwayRed.powerBar += power
                    if (runwayRed.powerBar >= 100) {
                        runwayRed.powerBar = 100
                    }
                }
                break;
            case 'green':
                if (!runwayGreen.bonusState) {
                    runwayGreen.powerBar += power
                    if (runwayGreen.powerBar >= 100) {
                        runwayGreen.powerBar = 100
                    }
                }
                break;
            case 'blue':
                if (!runwayBlue.bonusState) {
                    runwayBlue.powerBar += power
                    if (runwayBlue.powerBar >= 100) {
                        runwayBlue.powerBar = 100
                    }
                }
                break;
        }
    }

    updateSkillRating(value) {
        switch (value) {
            case 'gain':
                this.skillRating += 10
                break;
            case 'lose':
                this.skillRating -= 25
                break;
        }

        if (this.skillRating >= 100) {
            this.promoteRank()
        } else if (this.skillRating <= 0) {
            this.demoteRank()
        }
    }

    promoteRank() {
        if (this.gravityLvl >= 10) {
            this.skillRating = 100
        } else {
            this.gravityLvl++
            gravityLevelDisplay.style.animation = 'none'
            setTimeout(() => {
                gravityLevelDisplay.style.animation = '0.2s pulse'
            }, 0)
            this.updateGravityLevel()
            this.skillRating = 0
        }
    }

    demoteRank() {
        if (this.gravityLvl <= 1) {
            this.gravityLvl = 1
        } else {
            this.gravityLvl--
            this.updateGravityLevel()
        }
        this.skillRating = 0
    }

    updateGravityLevel() {
        canvas.style.transition = '1s'
        canvas.style.boxShadow = `0px 0px 100px rgba(255, 255, 255, ${this.gravityLvl / 15})`
        this.gravity = 4 + this.gravityLvl

        if (this.highestGravity < this.gravityLvl) {
            this.highestGravity = this.gravityLvl
        }

        gravityLevelDisplay.innerHTML = this.gravityLvl
    }

    endGame() {
        clearTimeout(this.diskSpawnerId)
        this.finishUpGame()
    }

    finishUpGame() {
        setTimeout(() => {
            if (diskAssembly.red.length === 0 &&
                diskAssembly.green.length === 0 &&
                diskAssembly.blue.length === 0 &&
                diskAssembly.hitDisks.length === 0 &&
                diskAssembly.missedDisks.length === 0
            ) {
                this.loadPostGame()
            } else {
                this.finishUpGame()
            }
        }, 1000)
    }

    loadPostGame() {
        gameManager.gameState = 'Post-Game'

        canvas.style.transition = '1s'
        canvas.style.boxShadow = `0px 0px 100px rgba(255, 255, 255, 0.1)`

        finalScoreDisplay.innerHTML = this.currentScore
        highestGravityDisplay.innerHTML = `Level ${this.highestGravity}`
        longestStreakDisplay.innerHTML = this.longestStreak
        disksHitDisplay.innerHTML = `${this.diskHitCount} out of ${this.disksSpawned} (${Math.floor((this.diskHitCount / this.disksSpawned) * 100)}%)`
        averageAccuracyDisplay.innerHTML = `${this.averageAccuracy}%`

        mainMenuDisplay.style.display = 'none'
        inGameDisplay.style.display = 'none'
        postGameDisplay.style.display = 'flex'
    }

    loadMainMenu() {
        gameManager.gameState = 'Main-Menu'

        canvas.style.transition = '1s'
        canvas.style.boxShadow = `0px 0px 100px rgba(255, 255, 255, 1)`

        inGameDisplay.style.display = 'none'
        postGameDisplay.style.display = 'none'
        mainMenuDisplay.style.display = 'flex'
    }
}